lens/sqlrepo/tstracker/tracker.go

```diff
@@ -0,0 +1,283 @@
+
+		winpostUnits := make([][]byte, 0, len(hdr.WinPoStProof))
+		for _, wpp := range hdr.WinPoStProof {
+			unit := make([]byte, 8, 8+len(wpp.ProofBytes))
+			binary.BigEndian.PutUint64(unit, uint64(wpp.PoStProof))
+			winpostUnits = append(
+				winpostUnits,
+				append(unit, wpp.ProofBytes...),
+			)
+		}
+		tsd.HeaderBlocks[i].WinpostTypesAndProof = winpostUnits
+
+		// FIXME - perhaps remove this coherence sanity check some day...
+		if i != 0 {
+
+			switch {
+
+			case !beaconArrsEqual(hdr.BeaconEntries, headers[0].BeaconEntries):
+				return nil, xerrors.Errorf("unexpected Beaconentries mismatch in tipset: %s(%d)=%s vs %s(0)=%s",
+					hdr.Cid().String(), i, hdr.BeaconEntries,
+					headers[0].Cid().String(), headers[0].BeaconEntries,
+				)
+
+			case !cidArrsEqual(hdr.Parents, headers[0].Parents):
+				return nil, xerrors.Errorf("unexpected Parents mismatch in tipset: %s(%d)=%s vs %s(0)=%s",
+					hdr.Cid().String(), i, hdr.Parents,
+					headers[0].Cid().String(), headers[0].Parents,
+				)
+
+			case hdr.ParentWeight.String() != headers[0].ParentWeight.String():
+				return nil, xerrors.Errorf("unexpected Weight mismatch in tipset: %s(%d)=%s vs %s(0)=%s",
+					hdr.Cid().String(), i, hdr.ParentWeight.String(),
+					headers[0].Cid().String(), headers[0].ParentWeight.String(),
+				)
+
+			case hdr.Height != headers[0].Height:
+				return nil, xerrors.Errorf("unexpected Epoch mismatch in tipset: %s(%d)=%d vs %s(0)=%d",
+					hdr.Cid().String(), i, hdr.Height,
+					headers[0].Cid().String(), headers[0].Height,
+				)
+
+			case hdr.ParentStateRoot.String() != headers[0].ParentStateRoot.String():
+				return nil, xerrors.Errorf("unexpected ParentStateroot mismatch in tipset: %s(%d)=%s vs %s(0)=%s",
+					hdr.Cid().String(), i, hdr.ParentStateRoot.String(),
+					headers[0].Cid().String(), headers[0].ParentStateRoot.String(),
+				)
+
+			case hdr.ParentMessageReceipts.String() != headers[0].ParentMessageReceipts.String():
+				return nil, xerrors.Errorf("unexpected ParentMessageReceipts mismatch in tipset: %s(%d)=%s vs %s(0)=%s",
+					hdr.Cid().String(), i, hdr.ParentMessageReceipts.String(),
+					headers[0].Cid().String(), headers[0].ParentMessageReceipts.String(),
+				)
+
+			case hdr.Timestamp != headers[0].Timestamp:
+				return nil, xerrors.Errorf("unexpected Timestamp mismatch in tipset: %s(%d)=%d vs %s(0)=%d",
+					hdr.Cid().String(), i, hdr.Timestamp,
+					headers[0].Cid().String(), headers[0].Timestamp,
+				)
+
+			case hdr.ParentBaseFee.String() != headers[0].ParentBaseFee.String():
+				return nil, xerrors.Errorf("unexpected BaseFee mismatch in tipset: %s(%d)=%s vs %s(0)=%s",
+					hdr.Cid().String(), i, hdr.ParentBaseFee.String(),
+					headers[0].Cid().String(), headers[0].ParentBaseFee.String(),
+				)
+			}
+		}
+	}
+
+	return tsd, nil
+}
+
+func cidArrsEqual(a, b []cid.Cid) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i].KeyString() != b[i].KeyString() {
```

Comment:
Why not `!a[i].Equals(b[i])`

Reply:
Because I wasn't thinking :) Adjusted in the other version of this code, thanks!